### PR TITLE
Minimalistic revisions for fixing distributed map file and master map management

### LINKED ISSUE #103: The mapfile banner should not be duplicated
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: 5.4
- Ruby: 2.1
- Distribution: RedHat 7.4
- Module version: 4.3.1

## How to reproduce (e.g Puppet code you use)
The issue revolves around defining the contents of a single map file via multiple `autofs::map` resources.  This is a useful thing to do, because it allows the mapfile definition to be distributed among different classes, and it already works.  But the current module implementation emits a warning banner into the target map file for **each** contributing `autofs::map` resource, instead of just one banner at the top of the file.

You don't even need to weigh the merits of distributed map file definitions, because the issue can be reproduced by applying a catalog built from as little as this:
```
include '::autofs'

autofs::mount { 'test':
  ensure  => 'present',
  mount   => '/mnt/test',
  mapfile => '/etc/test.auto'
}

autofs::map { 'test1':
  mapfile     => '/etc/test.auto',
  mapcontents => ['test1 -rw test.example.com:/export/test1']
}
```
(There is a workaround for that particular case, however.)
## What are you seeing
In the example manifest above, the `autofs::mount` declares an `autofs::map` with the (default; empty) `mapcontents` that is declared directly on the `autofs::mount`, so there are two `autofs::map` resources contributing to `/etc/test.auto`.  Result:
```
###############################################################
#                                                             #
# THIS FILE IS MANAGED BY PUPPET. ANY CHANGES MADE TO THIS    #
#   FILE WILL BE REVERTED BACK ON THE NEXT PUPPET RUN.        #
#                                                             #
###############################################################

###############################################################
#                                                             #
# THIS FILE IS MANAGED BY PUPPET. ANY CHANGES MADE TO THIS    #
#   FILE WILL BE REVERTED BACK ON THE NEXT PUPPET RUN.        #
#                                                             #
###############################################################

test1 -rw test.example.com:/export/test1
```
If there were more contributing `autofs::map` resources, there would be an additional banner for each.

The extra banners do not invalidate the map file, of course, but they do make it needlessly hard for a human to read.  It gets worse when there are more `autofs::map` resources, so that the actual mappings are sandwiched between banners.

## What behaviour did you expect instead
The banner ought to be presented only once, at the top of the map file, no matter how many `autofs::map` resources contribute to it:
```
###############################################################
#                                                             #
# THIS FILE IS MANAGED BY PUPPET. ANY CHANGES MADE TO THIS    #
#   FILE WILL BE REVERTED BACK ON THE NEXT PUPPET RUN.        #
#                                                             #
###############################################################

test1 -rw test.example.com:/export/test1
```
## Output log
(irrelevant)

## Any additional information you'd like to impart
It looks like this should be easy to fix.  The `concat` resource type has a `warn` parameter that serves exactly the purpose of presenting a warning banner, and it can be used to present a customized banner.  This (autofs) module does not use that, but it probably should.  Instead, it puts the banner -- a selection between two alternatives, in fact -- in the per-`autofs::map` templates.  Even if it were for some reason undesirable to use `$autofs::warn` for this purpose, the problem could also be solved by putting the banner in its own `concat::fragment`, and ensuring that that fragment comes first.

### LINKED ISSUE #104: Executable maps cannot be built from multiple pieces
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: 5.4.0
- Ruby: 2.1
- Distribution: RedHat Enterprise Linux 7.4
- Module version: Current HEAD of the git master branch (commit 4dac54d)

## How to reproduce (e.g Puppet code you use)
```
class { 'autofs': }
autofs::mount { 'data':
  mount   => '/mnt/data',
  mapfile => '/etc/auto.data',
  execute => true,  # this is the key attribute
}
autofs::map { 'datamapA':
  mapfile     => '/etc/auto.data',
  mapcontents => [ 'dataA -o rw /mnt/dataA', 'dataB -o rw /mnt/dataB' ],
}
```
## What are you seeing
Catalog building fails with a duplicate resource error.

## What behaviour did you expect instead
The catalog should be successfully built and applied.

## Output log
Error: Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Error while evaluating a Function Call, Duplicate declaration: Concat[/etc/auto.data] is already declared in file /opt/puppetlabs/puppet/modules/autofs/manifests/map.pp:54; cannot redeclare at /opt/puppetlabs/puppet/modules/autofs/manifests/map.pp:54 at /opt/puppetlabs/puppet/modules/autofs/manifests/map.pp:54:3  at /opt/puppetlabs/puppet/modules/autofs/manifests/mount.pp:161 on node centos-7-x64.xxx.yyy

## Any additional information you'd like to impart
If the `execute => true` attribute is not declared on `Autofs::Mount[data]` then the catalog compiles successfully and is applied with the expected result (I have a Beaker acceptance test demonstrating both the success case and the failure case).

### LINKED ISSUE #107: Catalog compilation can fail when managing the same map file via multiple autofs::map resources
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: 5.4.0
- Ruby: 2.1
- Distribution: RedHat Enterprise Linux 7.4
- Module version: 4.3.0, 4.3.1rc0

## How to reproduce (e.g Puppet code you use)
This is a generalization of issue #104. The underlying problem is that the module relies on `autofs::map` to declare an appropriate `Concat` resource for its map file, which it does indirectly by means of `ensure_resource()` in an attempt to avoid duplicate resource declarations when multiple `autofs::map` resources target the same map file.  But this strategy is only partially effective: it works only as long as all `autofs::map` resources contributing to the same map file pass _identical_ parameter lists to `ensure_resource()`.

The example in issue #104 demonstrates this problem, and so does this:
```
autofs::map { 'foo':
  mapfile => '/etc/auto.test',
  replace => false,
}

autofs::map { 'bar':
  mapfile => '/etc/auto.test',
  mapcontents => 'bar -rw example.com:/export/bar',
}
```
(That can also be rewritten to avoid the using the undocumented `$autofs::map::replace` parameter directly by using an `autofs::mount` resource with `$replace => false` in place of `Autofs::Map[foo]`.)

## What are you seeing
Catalog compilation fails with a duplicate resource declaration error.

## What behaviour did you expect instead
Either catalog compilation should succeed, somehow merging map file properties from multiple `autofs::map` declarations, or it should fail with a user-actionable error message (i.e. one describing a problem with the resources declared explicitly, not those declared internally by the module).

But really, the module design ought not to allow the problem to arise in the first place.  Or at bare minimum, the module documentation should present the relevant usage constraints.

## Output log
Error: Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Error while evaluating a Function Call, Duplicate declaration: Concat[/etc/auto.test] is already declared at (file: /home/jbolling/.puppetlabs/etc/code/modules/autofs/manifests/map.pp, line: 54); cannot redeclare (file: /home/jbolling/.puppetlabs/etc/code/modules/autofs/manifests/map.pp, line: 54) (file: /home/jbolling/.puppetlabs/etc/code/modules/autofs/manifests/map.pp, line: 54, column: 3) (file: /home/jbolling/tmp/map_err.pp, line: 7) on node xxx.yyy.org

## Any additional information you'd like to impart
I account the issue to a design flaw: defined type `autofs::map` is trying to do too many things.  It represents a chunk of contents of a map file, and with the possibility of multiple resources of that type contributing to the same map file, it is too much for `autofs::map` to have overall responsibility for the map file itself.  That ought to be the job of `autofs::mount`.

Making `autofs::mount` responsible for the overall map file (via that file's master `Concat` resource) would mean that one could not use `autofs::map` without `autofs::mount` (or equivalent data), but that's already halfway the case.  Without an `autofs::mount` to manage an appropriate entry in the master map (leaving aside the map file for a moment) the use cases for managing map files are relatively few.  And these could be addressed by giving `autofs::mount` the capability of managing _only_ the specified map file, leaving the master map alone.

### PR DESCRIPTION
**Pull request by jcbollinger**
#### Pull Request (PR) description
The changes in this pull request have two main thrusts:

 1. Enabling entries in the master map and overall map files to be managed independently of each other, and
 2. Eliminating opportunities for map file properties to be specified inconsistently, such as are the underlying cause of issues #104 and #107.

It attempts to accomplish this as much as possible within the framework already in place, via two key changes:

 1. Adding a `master_manage` property to `autofs::mount` to control whether that resource manages an entry in the master map, and

 2. Moving management of overall map file parameters (via the appropriate `Concat` resource) out of `autofs::map`, back to `autofs::mount`.

In addition to fixing several open issues, this has the welcome effect of making the odd behavior of `autofs::map::ensure` more consistent with other resources.

#### This Pull Request (PR) fixes the following issues
Fixes #103 
Fixes #104 
Fixes #107

### CODE REVIEW

File: manifests/map.pp [hunk 1]
```diff
@@ -11,61 +15,77 @@
...
   Enum['present', 'absent'] $ensure                                 = 'present',
   Stdlib::Absolutepath $mapfile                                     = $title,
-  Enum['autofs/auto.map.erb', 'autofs/auto.map.exec.erb'] $template = 'autofs/auto.map.erb',
-  String $mapmode                                                   = '0644',
-  Boolean $replace                                                  = true,
-  Integer $order                                                    = 1,
-  Variant[Array, String] $mapcontents                               = [],
+  Optional[String] $template                                        = undef,     # ignored
+  Optional[String] $mapmode                                         = undef,     # ignored
+  Optional[Boolean] $replace                                        = undef,     # ignored
```
Comment (dhollinger) [inline manifests/map.pp]: Making these three parameters ignored/no longer used makes the changes in this PR Backwards Incompatible.

Considering that both this and #119 would both be backwards incompatible merges, I'd rather take #119 than this one as they both would require new major versions anyway.

File: manifests/mount.pp [hunk 2]
```diff
@@ -20,51 +26,90 @@
...
-  Integer $order                          = 1,
+  Integer $order                          = 10,
   Optional[Variant[Stdlib::Absolutepath,Autofs::Mapentry]] $mapfile = undef,
   Optional[String] $options               = '',
   Stdlib::Absolutepath $master            = $autofs::auto_master_map,
+  Boolean $master_manage                  = true,
   Stdlib::Absolutepath $map_dir           = '/etc/auto.master.d',
   Boolean $use_dir                        = false,
-  Boolean $direct                         = true,
+  Optional[Boolean] $direct               = undef,  # deprecated
```
Comment (dhollinger) [inline manifests/mount.pp]: Same as previous comment.

**Comment by dhollinger** [comment]
I think at this point, with the changes being backwards-incompatible, I'd like to merge #119 instead of this.

@bastelfreak @ekohl @traylenator - any thoughts?

**Comment by jcbollinger** [comment]
Yes, both PRs are backwards-incompatible.  This is unavoidable because the key flaw underlying issues #104 and #109 is a _design_ flaw, introduced in commit d10bbda3.  At that point, management of the Concat resources for map files was moved from autofs::mount to autofs::map, producing an opportunity for inconsistency when there are two or more autofs::map instances for the same map file – a primary use case for that type.  When such inconsistencies occur, as they do in the examples in the issue reports, catalog building fails.  Thus, both of the PRs now under consideration pull that back out of autofs::map, one returning it to autofs::mount, and the other putting it in an altogether new type.

Both PRs being backwards incompatible, I, too, prefer #119, which I think produces a cleaner design than the alternative, with better separation of concerns.

**Comment by dhollinger** [comment]
Closing in favor of #119

**Pull request closed without merging**